website/docs/r/cloudfront_distribution.html.markdown

```diff
@@ -142,6 +142,100 @@ resource "aws_cloudfront_distribution" "s3_distribution" {
 }
 ```
 
+### CloudFront distribution with custom origin
+
+The example below uses an S3 static web hosting origin for use-cases beyond asset downloads. AWS doesn't allow using the S3 website endpoint from CloudFront. This example works using a custom authentication and a custom origin to further provide caching and HTTPS.
+
+```hcl
+locals {
+  origin_id = "S3-Website-${aws_s3_bucket.website-bucket.bucket}"
+  cloudfront-authentication-user-agent = "somethingRandomEen8Ohta"
+}
+
+resource "aws_s3_bucket" "website-bucket" {
+  bucket = "webhosting-bucket-via-cloudfront"
+
+  website {
+    index_document = "index.html"
+  }
+}
+
+resource "aws_s3_bucket_policy" "bucket-policy" {
+  bucket = "${aws_s3_bucket.website-bucket.id}"
+
+  policy = <<POLICY
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Sid": "ReadWithSecretUserAgent",
+            "Effect": "Allow",
+            "Principal": "*",
+            "Action": "s3:GetObject",
+            "Resource": "arn:aws:s3:::${aws_s3_bucket.website-bucket.id}/*",
+            "Condition": {
+                "StringEquals": {
+                    "aws:UserAgent": "${local.cloudfront-authentication-user-agent}"
+                }
+            }
+        }
+    ]
+}
+POLICY
+}
+
+resource "aws_cloudfront_distribution" "distribution" {
+  origin {
+    domain_name = "${aws_s3_bucket.website-bucket.website_endpoint}"
+    origin_id   = "${local.origin_id}"
+
+    custom_origin_config {
+      http_port              = "80"
+      https_port             = "443"
+      origin_protocol_policy = "http-only"
+      origin_ssl_protocols   = ["TLSv1", "TLSv1.1", "TLSv1.2"]
+    }
+
+    custom_header {
+      name  = "User-Agent"
+      value = "${local.cloudfront-authentication-user-agent}"
+    }
+  }
+
+  enabled = true
+  comment = "A distribution for our S3 Website Endpoint"
+
+  default_cache_behavior {
+    allowed_methods  = ["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
+    cached_methods   = ["HEAD", "GET", "OPTIONS"]
+    target_origin_id = "${local.origin_id}"
+
+    forwarded_values {
+      query_string = false
+
+      cookies {
+        forward = "all"
```

Comment:
you probably don't need to forward any cookies here, because s3 doesn't care if they are present. It will improve caching on Cloudfront. Same with most of the `headers`, except CORS ones.

Reply:
All tests were - of course - successful. Sorry for the delay. Changed the cookie forwarding to `none` in 754f0a88a3b0cee0439c2694b27cfe73c5c20e0c.

Reply:
@gersilex I actually meant this: https://www.terraform.io/docs/providers/aws/r/cloudfront_distribution.html#headers

However, I've checked that if you don't specify any value here, then it forwards no headers, which is good :)
So it's all good, sorry for confusion.